### PR TITLE
C#: Add implicit cast from array to pointer

### PR DESCRIPTION
**Pull request by tamasvajk**
During the testing of sign analysis we found that some expressions have no signs. One particular case was `int* p = ARR`, where the pointer is considered a numeric type, therefore we expect it to have a sign, but `ARR` is an array, which has no sign, so we end up with no sign for the pointer.

One way out of this issue would be to add an implicit cast from array to pointer. 

This PR at the moment only adds some test cases to see what are the extracted types.

### CODE REVIEW

File: csharp/ql/test/library-tests/conversion/pointer/Pointer.expected [hunk 1]
```diff
@@ -0,0 +1,11 @@
+| Pointer.cs:7:21:7:28 | Pointer.cs:7:21:7:28 | Int32* | Int32[] | access to parameter arr |
+| Pointer.cs:11:21:11:32 | Pointer.cs:11:21:11:32 | Int32* | Int32* | &... |
+| Pointer.cs:13:18:13:24 | Pointer.cs:13:18:13:24 | Int32* | Int32* | access to local variable i2 |
+| Pointer.cs:14:13:14:23 | Pointer.cs:14:13:14:23 | Int32* | Int32* | ... + ... |
+| Pointer.cs:15:13:15:25 | Pointer.cs:15:13:15:25 | Int32 | Int32 | ... + ... |
+| Pointer.cs:16:19:16:25 | Pointer.cs:16:19:16:25 | Void* | Void* | (...) ... |
```
Comment (tamasvajk) [inline csharp/ql/test/library-tests/conversion/pointer/Pointer.expected:6]: This is most probably an implicit cast from `int*` to `void*`.

File: csharp/ql/test/library-tests/conversion/pointer/Pointer.expected [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+| Pointer.cs:19:14:19:22 | Pointer.cs:19:14:19:22 | Int32* | null | null |
```
Comment (tamasvajk) [inline csharp/ql/test/library-tests/conversion/pointer/Pointer.expected:7]: According to [doc](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/unsafe-code-pointers/pointer-conversions), this should be an implicit cast.

File: csharp/ql/test/library-tests/conversion/pointer/Pointer.expected [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+| Pointer.cs:21:13:21:25 | Pointer.cs:21:13:21:25 | Int32 | Int32 | 1024 |
+| Pointer.cs:22:15:22:32 | Pointer.cs:22:15:22:32 | Byte* | Byte* | (...) ... |
+| Pointer.cs:24:13:24:29 | Pointer.cs:24:13:24:29 | String | String | "some string" |
+| Pointer.cs:25:22:25:27 | Pointer.cs:25:22:25:27 | Char* | String | access to local variable s |
```
Comment (tamasvajk) [inline csharp/ql/test/library-tests/conversion/pointer/Pointer.expected:11]: @hvitved The question is how to handle the cases on line 7 and 25.

**Review by tamasvajk — COMMENTED** [review]
@hvitved Let me know if you have some thoughts on how to proceed.

File: csharp/ql/test/experimental/ir/ir/raw_ir.expected [hunk 4]
```diff
@@ -1327,8 +1327,8 @@ pointers.cs:
 #    6|     r6_1(glval<Int32*>)  = VariableAddress[b]          : 
 #    6|     r6_2(glval<Int32[]>) = VariableAddress[arr]        : 
 #    6|     r6_3(Int32[])        = Load                        : &:r6_2, ~m?
-#    6|     r6_4(Int32*)         = Convert                     : r6_3
-#    6|     mu6_5(Int32*)        = Store                       : &:r6_1, r6_3
+#    6|     r6_4(Int32*)         = CheckedConvertOrThrow       : r6_3
```
Comment (tamasvajk) [inline csharp/ql/test/experimental/ir/ir/raw_ir.expected:1330]: I'm unsure what this change actually means. Or if it is expected at all.

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/Expressions/ImplicitCast.cs [hunk 5]
```diff
@@ -86,6 +86,15 @@ public static Expression Create(ExpressionNodeInfo info)
...
 
+            if (conversion.IsIdentity && conversion.IsImplicit &&
+                convertedType.Symbol is IPointerTypeSymbol &&
+                !(resolvedType.Symbol is IPointerTypeSymbol))
+            {
+                // int[] -> int*
+                // string -> char*
+                return new ImplicitCast(info);
+            }
+
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction.CSharp/Entities/Expressions/ImplicitCast.cs:97]: Looks very reasonable to me.

File: csharp/ql/test/experimental/ir/ir/raw_ir.expected (hunk 4 above)
Comment (hvitved) [reply csharp/ql/test/experimental/ir/ir/raw_ir.expected:1330]: I think it is fine; the IR would [insert a `Convert` operation](https://github.com/github/codeql/blob/main/csharp/ql/src/experimental/ir/implementation/raw/internal/TranslatedInitialization.qll#L142) when the RHS of an assignment did not match the LHS, but this is no longer needed. However, now there is an implicit cast, which translates to the `CheckedConvertOrThrow` operation.

**Comment by tamasvajk** [comment]
C# differences job: https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/459/

**Review by hvitved — APPROVED** [review]
LGTM, but let's wait for the differences job before merging.

**Comment by tamasvajk** [comment]
The differences job didn't reveal any issues. I'm merging this. cc @hvitved

**Pull request merged**